File: app/src/main/java/app/deadmc/devnetworktool/fragments/ping/PingChartPageFragment.java
package app.deadmc.devnetworktool.fragments.ping;

import android.graphics.Color;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.formatter.IFillFormatter;
import com.github.mikephil.charting.interfaces.dataprovider.LineDataProvider;
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;

import app.deadmc.devnetworktool.R;
import app.deadmc.devnetworktool.modules.PingStructure;

/**
 * Created by Feren on 16.01.2017.
 */
public class PingChartPageFragment extends BasePingFragment {

    private LineChart lineChart;
    private LineDataSet lineDataSet;

    public PingChartPageFragment() {

    }

    @Override
    public void addPingStructure(PingStructure pingStructure, boolean canUpdate){
        addDataToChart(pingStructure.getPing(),canUpdate);
    }

    @Override
    public void refreshFragment(ArrayList<PingStructure> pingStructureArrayList) {
        try {
            setPingStructureArrayList(pingStructureArrayList);
            lineChart.getData().notifyDataChanged();
            lineChart.notifyDataSetChanged();
            lineChart.invalidate();
        } catch (NullPointerException e) {}
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        setMyFragmentView(inflater.inflate(R.layout.fragment_pager_chart, container, false));
        restoreState(savedInstanceState);
        initElements();
        // Inflate the layout for this fragment
        return getMyFragmentView();
    }


    private void initElements() {
        initGraphic();
    }

    private void initGraphic() {

        lineChart = (LineChart) getMyFragmentView().findViewById(R.id.chart);
        lineChart.setViewPortOffsets(0, 0, 0, 0);
        lineChart.setBackgroundColor(Color.rgb(238, 238, 238));

        // no description text
        lineChart.getDescription().setEnabled(false);

        // enable touch gestures
        lineChart.setTouchEnabled(true);

        // enable scaling and dragging
        lineChart.setDragEnabled(true);
        lineChart.setScaleEnabled(true);

        // if disabled, scaling can be done on x- and y-axis separately
        lineChart.setPinchZoom(false);

        lineChart.setDrawGridBackground(false);
        lineChart.setMaxHighlightDistance(300);

        XAxis x = lineChart.getXAxis();
        x.setEnabled(false);

        YAxis y = lineChart.getAxisLeft();
        //y.setTypeface(mTfLight);
        y.setLabelCount(8, false);
        y.setTextColor(ContextCompat.getColor(getActivity(), R.color.textColor));
        y.setPosition(YAxis.YAxisLabelPosition.INSIDE_CHART);
        y.setDrawGridLines(false);
        y.setAxisLineColor(ContextCompat.getColor(getActivity(),R.color.textColor));

        lineChart.getAxisRight().setEnabled(false);

        // add data
        setDataToChart();

        lineChart.getLegend().setEnabled(false);

        lineChart.animateXY(2000, 2000);

        // dont forget to refresh the drawing
        lineChart.invalidate();
    }

    private void setDataToChart() {
        ArrayList<Entry> yVals = new ArrayList<Entry>();

        if (lineChart.getData() != null &&
                lineChart.getData().getDataSetCount() > 0) {
            lineDataSet = (LineDataSet) lineChart.getData().getDataSetByIndex(0);
            lineDataSet.setValues(yVals);
            lineChart.getData().notifyDataChanged();
            lineChart.notifyDataSetChanged();
        } else {
            // create a dataset and give it a type
            lineDataSet = new LineDataSet(yVals, "DataSet 1");

            lineDataSet.setMode(LineDataSet.Mode.HORIZONTAL_BEZIER);
            lineDataSet.setCubicIntensity(0.2f);
            lineDataSet.setDrawFilled(true);
            lineDataSet.setDrawCircles(false);
            lineDataSet.setLineWidth(1.0f);
            lineDataSet.setHighlightLineWidth(1.0f);
            lineDataSet.setHighLightColor(ContextCompat.getColor(getActivity(),R.color.textColor));
            lineDataSet.setColor(ContextCompat.getColor(getActivity(), R.color.colorPrimaryDark));
            lineDataSet.setFillColor(ContextCompat.getColor(getActivity(), R.color.colorPrimary));
            lineDataSet.setFillAlpha(150);
            lineDataSet.setDrawHorizontalHighlightIndicator(true);
            lineDataSet.setFillFormatter(new IFillFormatter() {
                @Override
                public float getFillLinePosition(ILineDataSet dataSet, LineDataProvider dataProvider) {
                    return 0;
                }
            });

            // create a data object with the datasets
            LineData data = new LineData(lineDataSet);
            //data.setValueTypeface(mTfLight);
            data.setValueTextSize(9f);
            data.setDrawValues(false);

            // set data
            lineChart.setData(data);
            if (getPingStructureArrayList() == null)
                return;
            Log.e("PingChart", "pingStructureArrayList size = " + getPingStructureArrayList().size());

            for(PingStructure pingStructure: getPingStructureArrayList()) {
                int i = lineDataSet.getEntryCount();
                lineDataSet.addEntry(new Entry(i, pingStructure.getPing()));
            }
            lineChart.getData().notifyDataChanged();
            lineChart.notifyDataSetChanged();
            lineChart.invalidate();
        }
    }

    private void addDataToChart(float value, boolean canUpdate) {
        int i = lineDataSet.getEntryCount();
        lineDataSet.addEntry(new Entry(i, value));
        if (canUpdate) {
            lineChart.getData().notifyDataChanged();
            lineChart.notifyDataSetChanged();
            lineChart.invalidate();
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        //Log.e("onSaveInstanceState", "save");
        outState.putString("pingStructureList", new Gson().toJson(getPingStructureArrayList()));
    }

    private void restoreState(Bundle bundle) {
        //Log.e("restoreState","restore");
        if (bundle==null)
            return;
        if (!bundle.containsKey("pingStructureList"))
            return;
        //Log.e("restoreState", "restore is ok");
        Type type = new TypeToken<ArrayList<PingStructure>>(){}.getType();
        String serializedString = bundle.getString("pingStructureList");
        Gson gson = new Gson();
        setPingStructureArrayList((ArrayList<PingStructure>)gson.fromJson(serializedString, type));
    }

}

File: app/src/main/java/app/deadmc/devnetworktool/fragments/ping/PingChartPageFragment.kt
package app.deadmc.devnetworktool.fragments.ping

import android.graphics.Color
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import app.deadmc.devnetworktool.R
import app.deadmc.devnetworktool.interfaces.views.PingView
import app.deadmc.devnetworktool.models.PingStructure
import app.deadmc.devnetworktool.presenters.BasePresenter
import app.deadmc.devnetworktool.presenters.PingPresenter
import com.arellomobile.mvp.presenter.InjectPresenter
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.formatter.IFillFormatter
import kotlinx.android.synthetic.main.fragment_pager_chart.view.*
import kotlinx.android.synthetic.main.layout_empty_list.view.*
import java.util.*


class PingChartPageFragment : PingBaseFragment(), PingView {

    @InjectPresenter
    lateinit var pingPresenter: PingPresenter
    private lateinit var lineDataSet: LineDataSet

    override fun addPingStructure(pingStructure: PingStructure) {
        if (initCompleted)
            addDataToChart(pingStructure.ping)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        myFragmentView = inflater!!.inflate(R.layout.fragment_pager_chart, container, false)
        initGraphic()
        initCompleted = true
        return myFragmentView
    }

    override fun getPresenter(): BasePresenter<*> {
        return pingPresenter
    }

    private fun initGraphic() {
        myFragmentView.chart.setViewPortOffsets(0f, 0f, 0f, 0f)
        myFragmentView.chart.setBackgroundColor(Color.rgb(238, 238, 238))
        // no description text
        myFragmentView.chart.description.isEnabled = false
        // enable touch gestures
        myFragmentView.chart.setTouchEnabled(true)
        // enable scaling and dragging
        myFragmentView.chart.isDragEnabled = true
        myFragmentView.chart.setScaleEnabled(true)
        // if disabled, scaling can be done on x- and y-axis separately
        myFragmentView.chart.setPinchZoom(false)
        myFragmentView.chart.setDrawGridBackground(false)
        myFragmentView.chart.maxHighlightDistance = 300f

        myFragmentView.chart.setBackgroundColor(ContextCompat.getColor(activity, R.color.white))

        val x = myFragmentView.chart.xAxis
        x.isEnabled = false
        val y = myFragmentView.chart.axisLeft
        //y.setTypeface(mTfLight);
        y.setLabelCount(8, false)
        y.axisMinimum = 0f
        y.textColor = ContextCompat.getColor(activity, R.color.textColor)
        y.setPosition(YAxis.YAxisLabelPosition.INSIDE_CHART)
        y.setDrawGridLines(false)
        y.axisLineColor = ContextCompat.getColor(activity, R.color.textColor)

        myFragmentView.chart.axisRight.isEnabled = false
        showEmpty()
        setDataToChart()
        myFragmentView.chart.legend.isEnabled = false
        myFragmentView.chart.animateXY(2000, 2000)
        myFragmentView.chart.invalidate()

    }

    private fun setDataToChart() {
        val yVals = ArrayList<Entry>()
        if (myFragmentView.chart.data != null && myFragmentView.chart.data.dataSetCount > 0) {
            showView()
            lineDataSet = myFragmentView.chart.data.getDataSetByIndex(0) as LineDataSet
            lineDataSet.values = yVals
            myFragmentView.chart.data.notifyDataChanged()
            myFragmentView.chart.notifyDataSetChanged()
        } else {
            // create a dataset and give it a type
            lineDataSet = LineDataSet(yVals, "DataSet 1")

            lineDataSet.mode = LineDataSet.Mode.HORIZONTAL_BEZIER
            lineDataSet.cubicIntensity = 0.2f
            lineDataSet.setDrawFilled(true)
            lineDataSet.setDrawCircles(false)
            lineDataSet.lineWidth = 1.0f
            lineDataSet.highlightLineWidth = 1.0f
            lineDataSet.highLightColor = ContextCompat.getColor(activity, R.color.textColorLight)
            lineDataSet.color = ContextCompat.getColor(activity, R.color.colorPrimaryDark)
            lineDataSet.fillColor = ContextCompat.getColor(activity, R.color.colorPrimary)
            lineDataSet.fillAlpha = 150
            lineDataSet.setDrawHorizontalHighlightIndicator(true)
            lineDataSet.fillFormatter = IFillFormatter { dataSet, dataProvider -> 0f }

            // create a data object with the datasets
            val data = LineData(lineDataSet!!)
            //data.setValueTypeface(mTfLight);
            data.setValueTextSize(9f)
            data.setDrawValues(false)
            myFragmentView.chart.data = data

            for (pingStructure in pingPresenter.pingStructureArrayList) {
                val i = lineDataSet.entryCount
                lineDataSet.addEntry(Entry(i.toFloat(), pingStructure.ping))
            }
            myFragmentView.chart.data.notifyDataChanged()
            myFragmentView.chart.notifyDataSetChanged()
            myFragmentView.chart.invalidate()
        }
    }

    private fun addDataToChart(value: Float) {
        showView()
        val i = lineDataSet.entryCount
        lineDataSet.addEntry(Entry(i.toFloat(), value))
        myFragmentView.chart.data.notifyDataChanged()
        myFragmentView.chart.notifyDataSetChanged()
        myFragmentView.chart.invalidate()
    }

    private fun showEmpty() {
        myFragmentView.chart.visibility = View.GONE
        myFragmentView.emptyLayout.visibility = View.VISIBLE
    }

    private fun showView() {
        if (myFragmentView.chart.visibility == View.VISIBLE)
            return
        myFragmentView.emptyLayout.visibility = View.GONE
        myFragmentView.chart.visibility = View.VISIBLE
    }


}
